doc/sg_timestamp.8
.TH SG_TIMESTAMP "8" "April 2018" "sg3_utils\-1.43" SG3_UTILS
.SH NAME
sg_timestamp \- report or set timestamp on SCSI device
.SH SYNOPSIS
.B sg_timestamp
[\fI\-\-elapsed\fR] [\fI\-\-help\fR] [\fI\-\-hex\fR]
[\fI\-\-milliseconds=MS\fR] [\fI\-\-no\-timestamp\fR] [\fI\-\-origin\fR]
[\fI\-\-raw\fR] [\fI\-\-readonly\fR] [\fI\-\-seconds=SECS\fR] [\fI\-\-srep\fR]
[\fI\-\-verbose\fR] [\fI\-\-version\fR] \fIDEVICE\fR
.SH DESCRIPTION
.\" Add any additional description here
Sends a SCSI REPORT TIMESTAMP or SET TIMESTAMP command to the \fIDEVICE\fR.
These commands are found in the SPC\-5 draft standard revision
7 (spc5r07.pdf).
.PP
If either the \fI\-\-milliseconds=MS\fR or \fI\-\-seconds=SECS\fR option is
given (and both can't be given) then the SET TIMESTAMP command is sent;
otherwise the REPORT TIMESTAMP command is sent.
.PP
The timestamp is sent and received from the \fIDEVICE\fR as the number of
milliseconds since the epoch of 1970\-01\-01 00:00:00 UTC and is held in a 48
bit unsigned integer. That same epoch is used by Unix machines, but they
usually hold the number of seconds since that epoch. The Unix date command
and especially its "+%s" format is useful in converting to and from
timestamps and more humanly readable forms. See the EXAMPLES section below.
.PP
A companion utility called sg_sat_datetime with similar options is available
to access or set the date and time in an ATA device (typically a SATA disk).
sg_sat_datetime issues ATA commands via the SCSI to ATA Translation (SAT)
layer.
.SH OPTIONS
Arguments to long options are mandatory for short options as well.
.TP
\fB\-e\fR, \fB\-\-elapsed\fR
assume the timestamp in the REPORT TIMESTAMP is an elapsed time from an
event such as a power cycle or hard reset and format the output as '<n>
days hh:mm:ss.xxx' where hh is hours (00 to 23 inclusive); mm is
minutes (00 to 59 inclusive); ss is seconds (00 to 59 inclusive) and xxx
is milliseconds (000 to 999 inclusive). If the number of days is 0
then '0 days' is not output unless this option is given two or more times.
.TP
\fB\-h\fR, \fB\-\-help\fR
output the usage message then exit.
.TP
\fB\-H\fR, \fB\-\-hex\fR
output the response to REPORT TIMESTAMP in ASCII hexadecimal on stderr. The
response is not decoded.
.TP
\fB\-m\fR, \fB\-\-milliseconds\fR=\fIMS\fR
where \fIMS\fR is the number of milliseconds since 1970\-01\-01 00:00:00 UTC
to set in the \fIDEVICE\fR with the SCSI SET TIMESTAMP command.
.TP
\fB\-N\fR, \fB\-\-no\-timestamp\fR
when REPORT TIMESTAMP is called this option suppress the output of the
timestamp value (in either seconds or milliseconds). This may be useful
in uncluttering the output when trying to decode the timestamp origin (see
the \fI\-\-origin\fR option).
.TP
\fB\-o\fR, \fB\-\-origin\fR
the REPORT TIMESTAMP returned parameter data contains a "timestamp origin"
field. When this option is given, that field is decoded and printed out
before the timestamp value is output. The default action (i.e. when the
option is not given) is not to print out this decoded field.
.br
T10 defines this field as "the most recent event that initialized the
returned device clock". The value 0 indicates a power up of hard reset
initialized the clock; 2 indicates a SET TIMESTAMP initialized the
clock while 3 indicates some other method initialized the clock.
.br
When used once a descriptive string is output (in a line before the
timestamp value). When used twice the value of the TIMESTAMP ORIGIN
field is output (in decimal, a value between 0 and 7 inclusive). When
used thrice a line of the form 'TIMESTAMP_ORIGIN=<value>' is output.
.TP
\fB\-r\fR, \fB\-\-raw\fR
output the SCSI REPORT TIMESTAMP response (i.e. the data\-out buffer) in
binary (to stdout). Note that the \fI\-\-origin\fR and \fI\-\-srep\fR
options are ignored when this option is given. Also all error and
verbose messages are output to stderr.
.TP
\fB\-R\fR, \fB\-\-readonly\fR
open the \fIDEVICE\fR read\-only. The default action is to open the
\fIDEVICE\fR read\-write.
.TP
\fB\-s\fR, \fB\-\-seconds\fR=\fISECS\fR
where \fISECS\fR is the number of seconds since 1970\-01\-01 00:00:00 UTC
to set in the \fIDEVICE\fR with the SCSI SET TIMESTAMP command. \fISECS\fR
is multiplied by 1000 before being used in the SET TIMESTAMP command.
.TP
\fB\-S\fR, \fB\-\-srep\fR
report the number of seconds since 1970\-01\-01 00:00:00 UTC. This is done
by dividing by 1000 the value returned by the SCSI REPORT TIMESTAMP command.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
increase the level of verbosity, (i.e. debug output).
.TP
\fB\-V\fR, \fB\-\-version\fR
print the version string and then exit.
.SH EXIT STATUS
The exit status of sg_timestamp is 0 when it is successful. Otherwise see
the sg3_utils(8) man page.
.SH NOTES
The TCMOS and the SCSIP bits in the Control extension mode page (see sdparm)
modify the actions of the timestamp held by a \fIDEVICE\fR.
.PP
Currently only the "Utilization usage rate based on date and time" parameters
within the Utilization log page (sbc4r09.pdf) use timestamps. See the sg_logs
utility. Vendor specific commands and pages may also be using timestamps.
.SH EXAMPLES
On Unix machines (e.g. Linux, FreeBSD and Solaris) the date command is useful
when working with timestamps.
.PP
To fetch the timestamp from a \fIDEVICE\fR and display it in a humanly
readable form the following could be used:
.PP
   # sg_timestamp \-S /dev/sdb
   1448993950
   # date \-\-date=@1448993950
   Tue Dec  1 13:19:10 EST 2015
   # date \-R \-\-date="@1448993950"
   Tue, 01 Dec 2015 13:19:10 \-0500
.PP
The latter two date commands show different forms of the same date (i.e.
1448993950 seconds since 1970\-01\-01 00:00:00 UTC). The sg_timestamp and
date commands can be combined using backquotes:
.PP
   # date \-R \-\-date=@`sg_timestamp \-S /dev/sdc`
   Wed, 16 Dec 2015 20:12:59 \-0500
.PP
To set the timestamp on the \fIDEVICE\fR to now (approximately) the
following could be used:
.PP
   # date +%s
   1448993955
   # sg_timestamp \-\-seconds=1448993955 /dev/sdb
.PP
Those two command lines could be combined into one by using backquotes:
.PP
   # sg_timestamp \-\-seconds=`date +%s` /dev/sdb
.SH AUTHORS
Written by Douglas Gilbert.
.SH "REPORTING BUGS"
Report bugs to <dgilbert at interlog dot com>.
.SH COPYRIGHT
Copyright \(co 2015\-2018 Douglas Gilbert
.br
This software is distributed under a BSD\-2\-Clause license. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B sdparm(sdparm), sg_logs(sg3_utils), sg_sat_datetime(sg3_utils)


CHANGELOG
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. All utilities in the main
directory have their own "man" pages. There is also a sg3_utils man page.

Changelog for sg3_utils-1.25 [20070710]
  - sg_dd: add oflag=sparse to step over bs*bpt number of zeros;
    - with oflag=sparse, write last bs*bpt segment at end or after
      error so file length of OFILE is appropriate
    - allow iflag=null and oflag=null which do nothing (placeholders)
    - when coe>1 then SCSI READ LONG logic remember extended block
      length of first encountered error
  - sg_ses: sync with ses2r17
  - place source in subversion repository

Changelog for sg3_utils-1.24 [20070507]
  - sg_raw: new utility to send arbirary SCSI commands
  - sg_luns: increase number of luns that can be fetched
    - fix length of raw and hex output
    - add '--quiet' option to output only ASCII hex
      representation of each lun
  - sg_rtpg: update for changes in spc4r09
  - sg_persist: update documentation, spc-4 references
    - fix exit status values
  - sg_inq: update version descriptors per spc4r09
    - fix '--id' and '--extended'
    - extend block limits VPD page (sbc3r09)
  - sg_vpd: extend block limits VPD page (sbc3r09)
    - append relative target port identifier to SAS target
      port address with '-iq' option
  - sg_logs: add decoding for stats+performance log pages
    - fix showing of page names for pdt > 0
    - implement '-HH' for single and all pages, fix '-r'
    - when '--maxlen=' given, only do single fetch
    - add Tape Alert (ssc), Media and Element statistics (smc) pages 
    - add '--brief' option
  - sg_ses: sync with ses2r16
    - fix bay number for SAS
  - sg_format: add '--dcrt' and '--security' options
  - sgm_dd: add 'smmap' oflag for shared_mmap_io testing
    - add 'dio' oflag
  - sg_dd, sgp_dd: add 'dio' iflag and oflag
  - sg_modes: change SAS mode page names per sas2r09
    - check validity of block descriptors length
  - sg_pt: change opaque context object from 'void *'
    to 'struct sg_pt_base *'
  - sg_opcodes: anticipate extra tmfs from 07-159r0
  - sg_sat_set_features: add more usage information
    - add man page
  - sg_sat_phy_event: add to examples directory
  - sg_lib: sync asc/ascq codes with spc4r10
  - Solaris port: using uscsi interface

Changelog for sg3_utils-1.23 [20070131]
  - sg_read_buffer: new utility
  - sg_write_buffer: new utility
  - sg_opcodes, sg_senddiag, sg_logs, sg_modes, sg_start, sg_inq,
    sg_turs, sg_readcap, sg_rbuf: add getopt_long() based cli;
    old and new cli selectable, new getopt_long cli is default
  - scripts: new subdirectory containing some bash scripts
    - add scripts/README file
  - sg_reassign: add '--hex' option for grown and primary lists
  - sg_rtpg: add '--raw' option
  - sg_lib.h, sg_cmds_basic.h + sg_cmds_extra.h: add C++ 
    'extern "C" ' wrappers
    - cleanup C code so it will compile as C++
  - sg_lib: sync with spc4r08
    - include <inttypes.h>, use PRId64 instead of %lld form
    - fix sg_get_sense_str() when empty sense buffer
  - win32 port: add Makefile.mingw + related support for MinGW
  - sg_cmds_extra: add sg_ll_read_buffer() and sg_ll_write_buffer()
  - sg_dd, sgp_dd, sgm_dd, sg_read: use lseek64() instead of llseek.c
  - sgm_dd: accept coe=<n> for interworking with sg_dd
  - sg_rdac: fix on non-linux ports
  - sg_ses: fix spurious warning in additional element status page
    - '-rr' option outputs a diagnostic page in binary to stdout
  - sg_opcodes: add command timeout descriptor support (spc4r08)
    - change linux specific pass through to generic pass through
  - sg_logs: add 'name=value' decoding for SAS specific lpage
  - examples+utils subdirectories: remove symlinks
  - synchronize man pages with usage messages
  - sg3_utils.spec: rework

Changelog for sg3_utils-1.22 [20061016]
  - sgp_dd: accept verbose=<n> as well as deb=<n> to ease
    interworking with sg_dd and sgm_dd
  - sg_sat_set_features: added to examples directory
  - sg_lib: sync asc/ascq text with spc4r06
    - move SG_LIB_CAT_NO_SENSE and SG_LIB_CAT_RECOVERED to
      20 and 21 respectively; add SG_LIB_CAT_ABORTED_COMMAND
      at 11 (its sense key value) 
  - sg_vpd: tweak '--page=sp --quiet' output
    - change '-HHH' so same as '-rr' (prepares ATA Information
      (ai) response for hdparm)
  - sg_requests: add '-s' option to set exit status from
    parameter data
  - sg_modes: exit quickly from '-e' if device not ready
  - sg_logs: sync sas log pages with sas2r05a
    - expand background scan results log page
    - add '-m=<max_len>' to limit response length
    - drop '-scum' and '-sthr' options and add '-select'
  - sg_write_long: add '--16' option to send 16 byte cdb
    - add '--wr_uncor' and '--pblock' options
  - sg_senddiag: cleanup and add sdiag_sas_p1_stop.txt
    to examples directory
  - sg_format: add '--cmplst=<n>' option (default: 1)
    - add '--pfu=<n>' option
    - expand man page to discuss P/D/C/GLISTs
  - sg_reassign: add '--primary' option to fetch primary
    defect list (PLIST) length
  - sg_readcap: add '-H' option to output response in hex
    and '-r' to output response in binary to stdout
    - add logical blocks per physical block (sbc3r07)
  - sginfo: add PLIST and GLIST designation to defect lists
  - sg_cmds: split this support file into sg_cmds_basic.[hc]
    and sg_cmds_extra.[hc]
    - add sg_ll_ata_pt() (SATL ATA pass) to sg_cmds_extra.[hc]
  - sg_rdac: fix includes for FreeBSD
  - sg_dd: add 'coe_limit=' option to exit after <n>
    consecutive 'coe' type read errors
  - sgm_dd: print out throughput information when signal arrives
    if time=1 (like sg_dd does already)
  - sg_inq: change '-HHH' so same as '-rr'. Now sg_inq, sg_vpd

src/sg_timestamp.c
/*
 * Copyright (c) 2015-2023 Douglas Gilbert.
 * All rights reserved.
 * Use of this source code is governed by a BSD-style
 * license that can be found in the BSD_LICENSE file.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <stdbool.h>
#include <string.h>
#include <ctype.h>
#include <getopt.h>
#define __STDC_FORMAT_MACROS 1
#include <inttypes.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "sg_lib.h"
#include "sg_lib_data.h"
#include "sg_pt.h"
#include "sg_cmds_basic.h"
#include "sg_unaligned.h"
#include "sg_pr2serr.h"

/* A utility program originally written for the Linux OS SCSI subsystem.
 *
 *
 * This program issues a SCSI REPORT TIMESTAMP and SET TIMESTAMP commands
 * to the given SCSI device. Based on spc5r07.pdf .
 */

static const char * version_str = "1.18 20230519";

static const char * my_name = "sg_timestamp: ";

#define REP_TIMESTAMP_CMDLEN 12
#define SET_TIMESTAMP_CMDLEN 12
#define REP_TIMESTAMP_SA 0xf
#define SET_TIMESTAMP_SA 0xf

#define SENSE_BUFF_LEN 64       /* Arbitrary, could be larger */
#define DEF_PT_TIMEOUT  60      /* 60 seconds */

uint8_t d_buff[256];

/* example Report timestamp parameter data */
/* uint8_t test[12] = {0, 0xa, 2, 0, 0x1, 0x51, 0x5b, 0xe2, 0xc1, 0x30,
 *                     0, 0}; */


static struct option long_options[] = {
        {"elapsed", no_argument, 0, 'e'},
        {"help", no_argument, 0, 'h'},
        {"hex", no_argument, 0, 'H'},
        {"milliseconds", required_argument, 0, 'm'},
        {"no_timestamp", no_argument, 0, 'N'},
        {"no-timestamp", no_argument, 0, 'N'},
        {"origin", no_argument, 0, 'o'},
        {"raw", no_argument, 0, 'r'},
        {"readonly", no_argument, 0, 'R'},
        {"seconds", required_argument, 0, 's'},
        {"srep", no_argument, 0, 'S'},
        {"verbose", no_argument, 0, 'v'},
        {"version", no_argument, 0, 'V'},
        {0, 0, 0, 0},
};

/* Indexed by 'timestamp origin' field value */
static const char * ts_origin_arr[] = {
    "initialized to zero at power on or by hard reset",
    "reserved [0x1]",
    "initialized by SET TIMESTAMP command",
    "initialized by other method",
    "reserved [0x4]",
    "reserved [0x5]",
    "reserved [0x6]",
    "reserved [0x7]",
};


static void
usage(int num)
{
    if (num > 1)
        goto page2;

    pr2serr("Usage: "
            "sg_timestamp  [--elapsed] [--help] [--hex] [--milliseconds=MS]\n"
            "                     [--no-timestamp] [--origin] [--raw] "
            "[--readonly]\n"
            "                     [--seconds=SECS] [--srep] [--verbose] "
            "[--version]\n"
            "                     DEVICE\n"
           );
    pr2serr("  where:\n"
            "    --elapsed|-e       show time as '<n> days hh:mm:ss.xxx' "
            "where\n"
            "                       '.xxx' is the remainder milliseconds. "
            "Don't show\n"
            "                       '<n> days' if <n> is 0 (unless '-e' "
            "given twice)\n"
            "    --help|-h          print out usage message, use twice for "
            "examples\n"
            "    --hex|-H           output response in ASCII hexadecimal\n"
            "    --milliseconds=MS|-m MS    set timestamp to MS "
            "milliseconds since\n"
            "                               1970-01-01 00:00:00 UTC\n"
            "    --no-timestamp|-N    suppress output of timestamp\n"
            "    --origin|-o        show Report timestamp origin "
            "(def: don't)\n"
            "                       used twice outputs value of field\n"
            "                       0: power up or hard reset; 2: SET "
            "TIMESTAMP\n"
            "    --raw|-r           output Report timestamp response to "
            "stdout in\n"
            "                       binary\n"
            "    --readonly|-R      open DEVICE read only (def: "
            "read/write)\n"
            "    --seconds=SECS|-s SECS    set timestamp to SECS "
            "seconds since\n"
            "                            1970-01-01 00:00:00 UTC\n"
            "    --srep|-S          output Report timestamp in seconds "
            "(def:\n"
            "                       milliseconds)\n"
            "    --verbose|-v       increase verbosity\n"
            "    --version|-V       print version string and exit\n\n"
          );
    pr2serr("Performs a SCSI REPORT TIMESTAMP or SET TIMESTAMP command. "
            "The timestamp\nis SET if either the --milliseconds=MS or "
            "--seconds=SECS option is given,\notherwise the existing "
            "timestamp is reported in milliseconds. The\nDEVICE stores "
            "the timestamp as the number of milliseconds since power up\n"
            "(or reset) or since 1970-01-01 00:00:00 UTC which also "
            "happens to\nbe the time 'epoch'of Unix machines.\n\n"
            "Use '-hh' (the '-h' option twice) for examples.\n"
#if 0
 "The 'date +%%s' command in "
            "Unix returns the number of\nseconds since the epoch. To "
            "convert a reported timestamp (in seconds since\nthe epoch) "
            "to a more readable form use "
            "'date --date=@<secs_since_epoch>' .\n"
#endif
           );
    return;
page2:
    pr2serr("sg_timestamp examples:\n\n"
            "It is possible that the target device containing a SCSI "
            "Logical Unit (LU)\nhas a battery (or supercapacitor) to "
            "keep its RTC (real time clock)\nticking during a power "
            "outage. More likely it doesn't and its RTC is\ncleared to "
            "zero after a power cycle or hard reset.\n\n"
            "Either way REPORT TIMESTAMP returns a 48 bit counter value "
            "whose unit is\na millisecond. A heuristic to determine if a "
            "date or elapsed time is\nbeing returned is to choose a date "
            "like 1 January 2000 which is 30 years\nafter the Unix epoch "
            "(946,684,800,000 milliseconds) and values less than\nthat are "
            "elapsed times and greater are timestamps. Observing the "
            "TIMESTAMP\nORIGIN field of REPORT TIMESTAMP is a better "
            "method:\n\n"
           );
    pr2serr(" $ sg_timestamp -o -N /dev/sg1\n"
            "Device clock initialized to zero at power on or by hard "
            "reset\n"
            " $ sg_timestamp -oo -N /dev/sg1\n"
            "0\n\n"
            " $ sg_timestamp /dev/sg1\n"
            "3984499\n"
            " $ sg_timestamp --elapsed /dev/sg1\n"
            "01:06:28.802\n\n"
            "The last output indicates an elapsed time of 1 hour, 6 minutes "
            "and 28.802\nseconds. Next set the clock to the current time:\n\n"
            " $ sg_timestamp --seconds=`date +%%s` /dev/sg1\n\n"
            " $ sg_timestamp -o -N /dev/sg1\n"
            "Device clock initialized by SET TIMESTAMP command\n\n"
            "Now show that as an elapsed time:\n\n"
            " $ sg_timestamp -e /dev/sg1\n"
            "17652 days 20:53:22.545\n\n"
            "That is over 48 years worth of days. Lets try again as a "
            "data-time\nstamp in UTC:\n\n"
            " $ date -u -R --date=@`sg_timestamp -S /dev/sg1`\n"
            "Tue, 01 May 2018 20:56:38 +0000\n"
           );
}

/* Invokes a SCSI REPORT TIMESTAMP command.  Return of 0 -> success,
 * various SG_LIB_CAT_* positive values or -1 -> other errors */
static int
sg_ll_rep_timestamp(int sg_fd, void * resp, int mx_resp_len, int * residp,
                    bool noisy, int verbose)
{
    int k, ret, res, sense_cat;
    uint8_t rt_cdb[REP_TIMESTAMP_CMDLEN] =
          {SG_MAINTENANCE_IN, REP_TIMESTAMP_SA, 0, 0,  0, 0, 0, 0, 0, 0, 0, 0};
    uint8_t sense_b[SENSE_BUFF_LEN] SG_C_CPP_ZERO_INIT;
    struct sg_pt_base * ptvp;

    sg_put_unaligned_be32((uint32_t)mx_resp_len, rt_cdb + 6);
    if (verbose) {
        char b[128];

        pr2serr("    Report timestamp cdb: %s\n",
                sg_get_command_str(rt_cdb, REP_TIMESTAMP_CMDLEN, false,
                                   sizeof(b), b));
    }

    ptvp = construct_scsi_pt_obj();
    if (NULL == ptvp) {
        pr2serr("%s: out of memory\n", __func__);
        return -1;
    }
    set_scsi_pt_cdb(ptvp, rt_cdb, sizeof(rt_cdb));
    set_scsi_pt_sense(ptvp, sense_b, sizeof(sense_b));
    set_scsi_pt_data_in(ptvp, (uint8_t *)resp, mx_resp_len);
    res = do_scsi_pt(ptvp, sg_fd, DEF_PT_TIMEOUT, verbose);
    ret = sg_cmds_process_resp(ptvp, "report timestamp", res, noisy, verbose,
                               &sense_cat);
    if (-1 == ret) {
        if (get_scsi_pt_transport_err(ptvp))
            ret = SG_LIB_TRANSPORT_ERROR;
        else
            ret = sg_convert_errno(get_scsi_pt_os_err(ptvp));
    } else if (-2 == ret) {
        switch (sense_cat) {
        case SG_LIB_CAT_RECOVERED:
        case SG_LIB_CAT_NO_SENSE:
            ret = 0;
            break;
        default:
            ret = sense_cat;
            break;
        }
    } else
        ret = 0;
    k = get_scsi_pt_resid(ptvp);
    if (residp)
        *residp = k;
    if ((verbose > 2) && ((mx_resp_len - k) > 0)) {
        pr2serr("Parameter data returned:\n");
        hex2stderr((const uint8_t *)resp, mx_resp_len - k,
                   ((verbose > 3) ? -1 : 1));
    }
    destruct_scsi_pt_obj(ptvp);
    return ret;
}

/* Invokes the SET TIMESTAMP command.  Return of 0 -> success, various
 * SG_LIB_CAT_* positive values or -1 -> other errors */
static int
sg_ll_set_timestamp(int sg_fd, void * paramp, int param_len, bool noisy,
                    int verbose)
{
    int ret, res, sense_cat;
    uint8_t st_cdb[SET_TIMESTAMP_CMDLEN] =
          {SG_MAINTENANCE_OUT, SET_TIMESTAMP_SA, 0, 0,  0, 0, 0, 0,
           0, 0, 0, 0};
    uint8_t sense_b[SENSE_BUFF_LEN] SG_C_CPP_ZERO_INIT;
    struct sg_pt_base * ptvp;

    sg_put_unaligned_be32(param_len, st_cdb + 6);
    if (verbose) {
        char b[128];

        pr2serr("    Set timestamp cdb: %s\n",
                sg_get_command_str(st_cdb, SET_TIMESTAMP_CMDLEN, false,
                                   sizeof(b), b));
        if ((verbose > 1) && paramp && param_len) {
            pr2serr("    set timestamp parameter list:\n");
            hex2stderr((const uint8_t *)paramp, param_len, -1);
        }
    }

    ptvp = construct_scsi_pt_obj();
    if (NULL == ptvp) {
        pr2serr("%s: out of memory\n", __func__);
        return -1;
    }
    set_scsi_pt_cdb(ptvp, st_cdb, sizeof(st_cdb));
    set_scsi_pt_sense(ptvp, sense_b, sizeof(sense_b));
    set_scsi_pt_data_out(ptvp, (uint8_t *)paramp, param_len);
    res = do_scsi_pt(ptvp, sg_fd, DEF_PT_TIMEOUT, verbose);
    ret = sg_cmds_process_resp(ptvp, "set timestamp", res, noisy, verbose,
                               &sense_cat);
    if (-1 == ret) {
        if (get_scsi_pt_transport_err(ptvp))
            ret = SG_LIB_TRANSPORT_ERROR;
        else
            ret = sg_convert_errno(get_scsi_pt_os_err(ptvp));
    } else if (-2 == ret) {
        switch (sense_cat) {
        case SG_LIB_CAT_RECOVERED:
        case SG_LIB_CAT_NO_SENSE:
            ret = 0;
            break;
        default:
            ret = sense_cat;
            break;
        }
    } else
        ret = 0;
    destruct_scsi_pt_obj(ptvp);
    return ret;
}

static void
dStrRaw(const uint8_t * str, int len)
{
    int k;

    for (k = 0; k < len; ++k)
        printf("%c", str[k]);
}


int
main(int argc, char * argv[])
{
    bool do_srep = false;
    bool do_raw = false;
    bool no_timestamp = false;
    bool readonly = false;
    bool secs_given = false;
    bool verbose_given = false;
    bool version_given = false;
    int res, c;
    int sg_fd = 1;
    int elapsed = 0;
    int do_origin = 0;
    int do_help = 0;
    int do_hex = 0;
    int do_set = 0;
    int ret = 0;
    int verbose = 0;
    uint64_t secs = 0;
    uint64_t msecs = 0;
    int64_t ll;
    const char * device_name = NULL;
    const char * cmd_name;

    if (getenv("SG3_UTILS_INVOCATION"))
        sg_rep_invocation(my_name, version_str, argc, argv, stderr);
    while (1) {
        int option_index = 0;

        c = getopt_long(argc, argv, "ehHm:NorRs:SvV", long_options,
                        &option_index);
        if (c == -1)
            break;

        switch (c) {
        case 'e':
            ++elapsed;
            break;
        case 'h':
        case '?':
            ++do_help;
            break;
        case 'H':
            ++do_hex;
            break;
        case 'm':
            ll = sg_get_llnum(optarg);
            if (-1 == ll) {
                pr2serr("bad argument to '--milliseconds=MS'\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            msecs = (uint64_t)ll;
            ++do_set;
            break;
        case 'N':
            no_timestamp = true;
            break;
        case 'o':
            ++do_origin;
            break;
        case 'r':
            do_raw = true;
            break;
        case 'R':
            readonly = true;
            break;
        case 's':
            ll = sg_get_llnum(optarg);
            if (-1 == ll) {
                pr2serr("bad argument to '--seconds=SECS'\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            secs = (uint64_t)ll;
            ++do_set;
            secs_given = true;
            break;
        case 'S':
            do_srep = true;
            break;
        case 'v':
            verbose_given = true;
            ++verbose;
            break;
        case 'V':
            version_given = true;
            break;
        default:
            pr2serr("unrecognised option code 0x%x ??\n", c);
            usage(1);
            return SG_LIB_SYNTAX_ERROR;
        }
    }
    if (optind < argc) {
        if (NULL == device_name) {
            device_name = argv[optind];
            ++optind;
        }
        if (optind < argc) {
            for (; optind < argc; ++optind)
                pr2serr("Unexpected extra argument: %s\n", argv[optind]);
            usage(1);
            return SG_LIB_SYNTAX_ERROR;
        }
    }
    if (do_help) {
        usage(do_help);
        return 0;
    }

#ifdef DEBUG
    pr2serr("In DEBUG mode, ");
    if (verbose_given && version_given) {
        pr2serr("but override: '-vV' given, zero verbose and continue\n");
        /* verbose_given = false; */
        version_given = false;
        verbose = 0;
    } else if (! verbose_given) {
        pr2serr("set '-vv'\n");
        verbose = 2;
    } else
        pr2serr("keep verbose=%d\n", verbose);
#else
    if (verbose_given && version_given)
        pr2serr("Not in DEBUG mode, so '-vV' has no special action\n");
#endif
    if (version_given) {
        pr2serr("version: %s\n", version_str);
        return 0;
    }

    if (do_set > 1) {
        pr2serr("either --milliseconds=MS or --seconds=SECS may be given, "
                "not both\n");
        usage(1);
        return SG_LIB_CONTRADICT;
    }

    if (NULL == device_name) {
        pr2serr("missing device name!\n\n");
        usage(1);
        return SG_LIB_SYNTAX_ERROR;
    }

    sg_fd = sg_cmds_open_device(device_name, readonly, verbose);
    if (sg_fd < 0) {
        if (verbose)
            pr2serr("open error: %s: %s\n", device_name,
                    safe_strerror(-sg_fd));
        ret = sg_convert_errno(-sg_fd);
        goto fini;
    }

    memset(d_buff, 0, 12);
    if (do_set) {
        cmd_name = "Set timestamp";
        sg_put_unaligned_be48(secs_given ? (secs * 1000) : msecs, d_buff + 4);
        res = sg_ll_set_timestamp(sg_fd, d_buff, 12, true, verbose);
    } else {
        cmd_name = "Report timestamp";
        res = sg_ll_rep_timestamp(sg_fd, d_buff, 12, NULL, true, verbose);
        if (0 == res) {
            if (do_raw)
                dStrRaw(d_buff, 12);
            else if (do_hex)
                hex2stderr(d_buff, 12, 1);
            else {
                int len = sg_get_unaligned_be16(d_buff + 0);

                if (len < 8)
                    pr2serr("timestamp parameter data length too short, "
                            "expect >= 10, got %d\n", len + 2);
                else {
                    if (do_origin) {
                        if (1 == do_origin)
                            printf("Device clock %s\n",
                                   ts_origin_arr[0x7 & d_buff[2]]);
                        else if (2 == do_origin)
                            printf("%d\n", 0x7 & d_buff[2]);
                        else
                            printf("TIMESTAMP_ORIGIN=%d\n", 0x7 & d_buff[2]);
                    }
                    if (! no_timestamp) {
                        msecs = sg_get_unaligned_be48(d_buff + 4);
                        if (elapsed) {
                            int days = (int)(msecs / 1000 / 60 / 60 / 24);
                            int hours = (int)(msecs / 1000 / 60 / 60 % 24);
                            int mins = (int)(msecs / 1000 / 60 % 60);
                            int secs_in_min =(int)( msecs / 1000 % 60);
                            int rem_msecs = (int)(msecs % 1000);

                            if ((elapsed > 1) || (days > 0))
                                printf("%d day%s ", days,
                                       ((1 == days) ? "" : "s"));
                            printf("%02d:%02d:%02d.%03d\n", hours, mins,
                                   secs_in_min, rem_msecs);
                        } else
                            printf("%" PRIu64 "\n", do_srep ?
                                                    (msecs / 1000) : msecs);
                    }
                }
            }
        }
    }
    ret = res;
    if (res) {
        if (SG_LIB_CAT_INVALID_OP == res)
            pr2serr("%s command not supported\n", cmd_name);
        else {
            char b[80];

            sg_get_category_sense_str(res, sizeof(b), b, verbose);
            pr2serr("%s command: %s\n", cmd_name, b);
        }
    }

fini:
    if (sg_fd >= 0) {
        res = sg_cmds_close_device(sg_fd);
        if (res < 0) {
            pr2serr("close error: %s\n", safe_strerror(-res));
            if (0 == ret)
                ret = sg_convert_errno(-res);
        }
    }
    if (0 == verbose) {
        if (! sg_if_can2stderr("sg_timestamp failed: ", ret))
            pr2serr("Some error occurred, try again with '-v' "
                    "or '-vv' for more information\n");
    }
    return (ret >= 0) ? ret : SG_LIB_CAT_OTHER;
}
